lib/index.js

```diff
@@ -0,0 +1,220 @@
+#! /usr/bin/env node
+'use strict';
+var esprima = require('esprima');
+var fs = require('fs');
+var utils = require('./utils.js');
+var walker = require('./walker.js');
+var promise = require('./promise.js');
+var callback = require('./callback.js');
+var conditional = require('./conditional.js');
+
+function generate_output_string(conditionals, global_exits, promises, callbacks){
+   /*
+      Generates final program output as a readable string.
+
+      @param conditionals: a sorted object of conditionals with a property for those that exit, and a property for those that do not
+      @param global_exits: an array of esprima nodes of program scope which exit
+      @param promises: an array of fcn calls that return promises, which have 
+         been parsed to identify when they exit
+      @param callbacks: an array of functions with their callbacks, which have 
+         been parsed to identify when they exit
+
+      @return: a string explaining problems or lack thereof.
+         - strings which begin 'Hooray!' always exit
+         - strings which begin 'For promised function' deal with promises
+         - strings which begin 'For callback chain' deal with callbacks
+         - strings which begin with none of the above never exit
+   */
+   if (global_exits.length){
+      return "Hooray! Your code contains a global exit.";
+   } else{
+      var cond_str = "";
+      var prom_str = "";
+      var call_str = "";
+      if (conditionals){
+         cond_str = conditional.output(conditionals);
+         if (/^Your code never exits/.exec(cond_str)){
+            // code never exits, we want to pretend there was no message
+            cond_str = "";
+         }
+      }
+      if (promises){
+         prom_str = promise.output(promises);
+         // console.log("prom_str");
+         if (/^Your code never exits/.exec(prom_str)){
+            // code never exits, we want to pretend there was no message
+            prom_str = "";
+         }
+      }
+      if (callbacks){
+         call_str = callback.output(callbacks);
+         if (/^Your code never exits/.exec(call_str)){
+            // code never exits, we want to pretend there was no message
+            call_str = "";
+         }
+      }
+      if (cond_str || prom_str || call_str){
+         if (/^Hooray!/.exec(cond_str) && (prom_str || call_str)){
```

Comment:
I don't think we should be testing our own strings to see if they exited. This is all internal right? Why not just pass around an object that contains the information?